proxy/proxy.go

```diff
@@ -0,0 +1,668 @@
+	return err
+}
+
+// Addr returns the listener's network address.
+func (l *TappedListener) Addr() net.Addr {
+	return l.proxyServer.tlsListener.Addr()
+}
+
+type ProxyServer struct {
+	ctx          context.Context
+	tlsListener  net.Listener
+	overflow     chan<- acceptResult
+	proxyHandler *ProxyHandler
+	baseConfig   *tls.Config
+}
+
+// NewTlsProxyServer returns a ProxyServer, which listens for TLS connections and will handle each connection either by
+// proxying to a running task's container or by passing the connection on to the associated `TappedListener`. This
+// decision is based on the SNI (Server Name Indicator) in the TLS Client Hello message. Connections that are made to
+// subdomains of `hostname` will be proxied to a container of the shard that's indicated by the subdomain labels.
+// Connections that are made to `hostname` exactly will be given to the returned `TappedListener` to be handled by
+// another server.
+func NewTlsProxyServer(hostname string, port uint16, tlsCerts []tls.Certificate, shardClient pc.ShardClient, jwtVerificationKey []byte) (*ProxyServer, *TappedListener, error) {
+	var tcpListener, err = net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		return nil, nil, err
+	}
+	var ctx, cancelFunc = context.WithCancel(context.Background())
+	var proxyHandler = newHandler(hostname, shardClient, jwtVerificationKey)
+	var tlsConfig = getTlsConfig(tlsCerts, proxyHandler)
+
+	var tlsListener = tls.NewListener(tcpListener, tlsConfig)
+
+	var acceptCh = make(chan acceptResult)
+	var server = &ProxyServer{
+		ctx:          ctx,
+		tlsListener:  tlsListener,
+		overflow:     acceptCh,
+		proxyHandler: proxyHandler,
+		baseConfig:   tlsConfig,
+	}
+	var tappedListener = &TappedListener{
+		cancelFunc:  cancelFunc,
+		proxyServer: server,
+		recv:        acceptCh,
+	}
+	return server, tappedListener, nil
+}
+
+// Run starts the server accepting and handling new connections, and blocks until
+// the server stops with an error (which will never be nil).
+func (ps *ProxyServer) Run() error {
+	var err error
+	defer func() {
+		if err == nil {
+			panic("ProxyServer.Run has nil terminal error")
+		}
+		log.Info("proxy server shutting down, sending final error to overflow listener")
+		select {
+		case ps.overflow <- acceptResult{err: err}:
+		case <-ps.ctx.Done():
+		}
+		close(ps.overflow)
+		log.Info("proxy server shutdown complete")
+	}()
+	for {
+		if err = ps.ctx.Err(); err != nil {
+			return err
+		}
+		var conn net.Conn
+		conn, err = ps.tlsListener.Accept()
+		if err != nil {
+			log.WithField("error", err).Error("failed to accept tls connection")
+			return err
+		}
+		// Start a new goroutine to hanlde this connection, so we don't block the accept loop
+		go func() {
+			// Await the completion of the TLS handshake. This is needed in order to
+			// ensure that ConnectionState is populated with the SNI from the client hello.
+			if hsErr := conn.(*tls.Conn).Handshake(); hsErr != nil {
+				// This may
```

Comment:
:facepalm: this was just going to be a not about verbosity of this log statement, since this one may get very spammy. I just removed the comment and changed the log level to Info for now.